community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/CheckPointerImpl.java

```diff
@@ -0,0 +1,158 @@
+import org.neo4j.kernel.impl.transaction.log.WritableLogChannel;
+import org.neo4j.kernel.impl.transaction.log.entry.LogEntryWriterV1;
+import org.neo4j.kernel.impl.transaction.log.pruning.LogPruning;
+import org.neo4j.kernel.impl.transaction.log.rotation.StoreFlusher;
+import org.neo4j.kernel.impl.transaction.tracing.CheckPointTracer;
+import org.neo4j.kernel.impl.transaction.tracing.LogCheckPointEvent;
+import org.neo4j.kernel.lifecycle.LifecycleAdapter;
+import org.neo4j.logging.Log;
+import org.neo4j.logging.LogProvider;
+
+import static org.neo4j.kernel.impl.transaction.log.checkpoint.CheckPointer.PrintFormat.prefix;
+
+public class CheckPointerImpl extends LifecycleAdapter implements CheckPointer
+{
+    private final LogFile logFile;
+    private final TransactionIdStore transactionIdStore;
+    private final CheckPointThreshold threshold;
+    private final StoreFlusher storeFlusher;
+    private final LogPruning logPruning;
+    private final KernelHealth kernelHealth;
+    private final Log msgLog;
+    private final CheckPointTracer tracer;
+
+    private WritableLogChannel channel;
+    private TransactionLogWriter transactionLogWriter;
+
+    public CheckPointerImpl( TransactionIdStore transactionIdStore,
+            CheckPointThreshold threshold, StoreFlusher storeFlusher, LogPruning logPruning, KernelHealth kernelHealth,
+            LogProvider logProvider, CheckPointTracer tracer, LogFile logFile )
+    {
+        this.logFile = logFile;
+        this.transactionIdStore = transactionIdStore;
+        this.threshold = threshold;
+        this.storeFlusher = storeFlusher;
+        this.logPruning = logPruning;
+        this.kernelHealth = kernelHealth;
+        this.msgLog = logProvider.getLog( CheckPointerImpl.class );
+        this.tracer = tracer;
+    }
+
+    @Override
+    public void start() throws Throwable
+    {
+        this.channel = logFile.getWriter();
+        this.transactionLogWriter =
+                new TransactionLogWriter( new LogEntryWriterV1( channel, NeoCommandHandler.EMPTY ) );
+    }
+
+    @Override
+    public void forceCheckPoint() throws IOException
+    {
+        synchronized ( logFile )
+        {
+            doCheckPoint();
+        }
+    }
+
+    @Override
+    public void checkPointIfNeeded()
+    {
+        if ( threshold.isCheckPointingNeeded() )
+        {
+            synchronized ( logFile )
+            {
+                if ( threshold.isCheckPointingNeeded() )
+                {
+                    try ( LogCheckPointEvent event = tracer.beginCheckPoint() )
+                    {
+                        try
+                        {
+                            doCheckPoint();
+                        }
+                        catch ( IOException ex )
+                        {
+                            throw new UnderlyingStorageException( ex );
+                        }
+                    }
+                }
+            }
```

Comment:
This is not good. When you synchronise on the logFile, you also prevent transactions from committing. You should only lock on it for writing out the check-point entries, and perhaps for reading the LogPosition of the last commit. Use a different strategy for ensuring that only one check-point is in progress at a time. You could for instance have an AtomicBoolean, where you only call doCheckPoint if you can successfully compareAndSet it to true, for instance.